### ISSUE TITLE
length mismatch for istft

### ISSUE THREAD

**Issue by nateanl** [pytorch/pytorch#62747]
## 🐛 Bug

As discussed in https://github.com/pytorch/audio/issues/1409
## To Reproduce

1. download stft.pt from [link](https://drive.google.com/file/d/1TjwwGFEotsjAay0osCqw1IlcuANX675R/view?usp=sharing)
1. run the following code

```
import torch

fname = 'stft.pt'
stft_feat = torch.load(fname)
print("stft_feat", stft_feat.size())

n_fft = 512
hop_length = 320
win_length = 512
window = torch.hann_window(win_length).cuda()
center = True
length = 150079
reconstruct = torch.istft(stft_feat, n_fft, hop_length=hop_length, win_length=win_length, window=window, center=center, length=length)[0]
print("reconstruct", reconstruct.size())
```

## Expected behavior

In my environment, the program printed

**stft_feat torch.Size([1, 257, 469])
reconstruct torch.Size([150016])**

The length seems to be incorrect. The original waveform has length of 150079.

## Reason
The reason is that the hop size is 320 that is a little big when n_fft is 512.
See the implementation of [`torch.istft`](https://github.com/pytorch/pytorch/blob/b6024b9d12b389faada6643dd21efa194a07c231/aten/src/ATen/native/SpectralOps.cpp#L923):
```
  const auto start = center ? n_fft / 2 : 0;
  const auto end = lengthOpt.has_value()? start + lengthOpt.value() : - n_fft / 2;
```
The start is set to 256 since center is True in your case.
The end is set to start + lengthOpt.value() that is 256+150079 = 150335. 
After slicing, the shape should be 150079, looks perfect. 

However, the [expected_signal_length](https://github.com/pytorch/pytorch/blob/b6024b9d12b389faada6643dd21efa194a07c231/aten/src/ATen/native/SpectralOps.cpp#L819) is `n_fft + hop_length * (n_frames - 1)` that is 512+320*468 = 150272. It's smaller than 150335
After slicing operation [256:150272], the shape of your waveform becomes 150016.

In librosa, they added a [padding logic](https://github.com/librosa/librosa/blob/main/librosa/core/spectrum.py#L436) if the expected length is shorter than the given length. 

To make torch.istft consistent with librosa, do we want to add such padding (fix_length) function in the code?

cc @mruberry @rgommers @heitorschueroff @peterbell10 @walterddr

**Comment by mruberry** [comment]
cc @peterbell10 -- what do you think? These operations are derived from librosa

We could even think about moving them to a librosa namespace: torch.librosa.stft, for example

**Comment by nateanl** [comment]
I'd like to volunteer adding a fix_length logic in istft and create a PR for that. 

In terms of moving istft to librosa namespace, I doubt it'll be BC-Breaking and affect users' current codes.

**Comment by peterbell10** [comment]
So the `istft` signal prior to slicing has shape `n_fft + hop_length * (length // hop_length)` and then the minimum lost to slicing is `n_fft // 2`. That means any `hop_length > n_fft // 2` can cause this issue depending on if the input signal evenly divides `hop_length` or not.

I think `librosa`'s approach of padding the signal is good here, but it would be good to add a note in the docstring that `hop_length > n_fft // 2`  may lead to issues.

**Comment by nateanl** [comment]
Resolved by https://github.com/pytorch/pytorch/pull/63469
